atc/db/build.go

```diff
@@ -549,6 +549,26 @@ func (b *build) Finish(status BuildStatus) error {
 		if err != nil {
 			return err
 		}
+
+		pipelineRows, err := pipelinesQuery.
```

Comment:
Nice! I think having pipelines archive immediately makes sense because it won't have the user guessing as to when the expected archival will happen (also will feel much more responsive).

One thing I wanted to get your opinion on was the fact that this will only archive the pipeline that is the direct child of the build/job that is finished. So for example, let's say pipeline-1 has a job that sets pipeline-2 and pipeline-2 has a job that sets pipeline-3. If the build of pipeline-1 archives pipeline-2, pipeline-3 will not be archived until the gc tick. This would result in a user seeing their direct child pipeline being archived immediately, but any pipelines that were set by that archived pipeline would have to wait 10 seconds (or more depending on how nested the pipelines are).

I'm not sure if you think this is something that we want to fix? I think for now with pipelines it might not be a big use case because people probably won't have more than 1 level deep of nested pipelines? It's just something that would be possibly weird to the user if they do end up running into this situation.

If you do think it's something you want to explore, I wrote up a query that would cover that case. It recursively finds any pipelines that are children of any pipelines we found that need to be archived.
```
WITH RECURSIVE pipelines_to_archive AS (
	SELECT id from pipelines where archived = false AND parent_job_id = ? AND parent_build_id < ?
	UNION
	SELECT p.id from pipelines p join jobs j on p.parent_job_id = j.id join pipelines_to_archive on j.pipeline_id = pipelines_to_archive.id)
SELECT * from pipelines_to_archive;
```

Reply:
Woah, didn't know WITH RECURSIVE was a thing

Reply:
DAMN that's cool! We did briefly talk about that exact example you gave and thought "nahh who would do recursive pipelines?!". We'll give it a shot to see if we can implement it since you've done the hard work of making the query.

Reply:
Haha ya that's kinda where I was at with it too, I'm not sure who would do pipelines setting other pipelines?? But I guess if it's a light change that won't cost much then it might to worth to add?